hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -688,6 +663,48 @@ public OmKeyInfo lookupKey(OmKeyArgs args, String clientAddress)
     }
   }
 
+  /**
+   * Refresh pipeline info in OM by asking SCM.
+   * @param value OmKeyInfo
+   */
+  @VisibleForTesting
+  protected void refreshPipeline(OmKeyInfo value) {
+    // Refresh container pipeline info from SCM
+    // based on OmKeyArgs.refreshPipeline flag
+    // 1. Client send initial read request OmKeyArgs.refreshPipeline = false
+    // and uses the pipeline cached in OM to access datanode
+    // 2. If succeeded, done.
+    // 3. If failed due to pipeline does not exist or invalid pipeline state
+    //    exception, client should retry lookupKey with
+    //    OmKeyArgs.refreshPipeline = true
+    Map<Long, ContainerWithPipeline> containerWithPipelineMap = new HashMap<>();
+    for (OmKeyLocationInfoGroup key : value.getKeyLocationVersions()) {
+      key.getLocationList().forEach(k -> {
+        // TODO: fix Some tests that may not initialize container client
+        // The production should always have containerClient initialized.
+        if (scmClient.getContainerClient() != null) {
+          ContainerWithPipeline cp = containerWithPipelineMap
+              .computeIfAbsent(k.getContainerID(), cId -> {
+                try {
+                  return scmClient.getContainerClient()
+                      .getContainerWithPipeline(cId);
+                } catch (IOException e) {
+                  LOG.error("Unable to update pipeline for container:{}",
+                      k.getContainerID());
+                  return null;
```

Comment:
Do we we need to fail here, instead of logging as error and return null. As when reading this will eventually fail right, as no pipeline info is available for that container.